File: Cpp/DataStructuresAndAlgorithms/GRAPHS.cpp
#include <iostream>
#include <map>
#include <list>
#include <queue>
#include <vector>

using namespace std;

// ==========================================
// GRAPH CLASS DEFINITION
// ==========================================
class Graph {
private:
    // Adjacency List:
    // Map Key (int) = The Vertex
    // Map Value (list<int>) = A list of all vertices connected to it
    map<int, list<int>> adj;

    // Helper function for DFS (Recursive)
    void dfsHelper(int node, map<int, bool> &visited) {
        // Step 1: Mark the current node as visited and print it
        visited[node] = true;
        cout << node << " ";

        // Step 2: Recur for all the vertices adjacent to this vertex
        for (int neighbor : adj[node]) {
            // If the neighbor hasn't been visited yet, go deeper!
            if (!visited[neighbor]) {
                dfsHelper(neighbor, visited);
            }
        }
    }

public:
    // --- CORE OPERATIONS ---

    // 1. ADD EDGE: Connects vertex 'u' and vertex 'v'
    // 'bidirectional' = true means if A is connected to B, B is also connected to A.
    void addEdge(int u, int v, bool bidirectional = true) {
        adj[u].push_back(v); // Connect u -> v
        
        if (bidirectional) {
            adj[v].push_back(u); // Connect v -> u
        }
    }

    // 2. BFS (Breadth-First Search) - The "Layer by Layer" Approach
    void bfs(int startNode) {
        cout << "BFS Traversal: ";
        
        // Use a map to keep track of who we have already visited
        map<int, bool> visited;
        
        // Use a QUEUE for BFS (First In, First Out)
        queue<int> q;

        // Step 1: Start with the initial node
        q.push(startNode);
        visited[startNode] = true;

        while (!q.empty()) {
            // Step 2: Get the node at the front
            int currentNode = q.front();
            q.pop();
            cout << currentNode << " ";

            // Step 3: Look at all neighbors of the current node
            for (int neighbor : adj[currentNode]) {
                // If we haven't visited this neighbor yet...
                if (!visited[neighbor]) {
                    q.push(neighbor); // Add to queue to visit later
                    visited[neighbor] = true; // Mark as visited so we don't add it twice
                }
            }
        }
        cout << endl;
    }

    // 3. DFS (Depth-First Search) - The "Deep Dive" Approach
    void dfs(int startNode) {
        cout << "DFS Traversal: ";
        
        // We need a visited map to stop infinite loops in cycles
        map<int, bool> visited;
        
        // Call the recursive helper
        dfsHelper(startNode, visited);
        
        cout << endl;
    }

    // 4. DISPLAY THE GRAPH
    void printGraph() {
        cout << "\n--- Adjacency List Structure ---" << endl;
        // Iterate over every key-value pair in the map
        for (auto row : adj) {
            cout << row.first << " -> "; // The Node
            for (int neighbor : row.second) { // Its Neighbors
                cout << neighbor << " ";
            }
            cout << endl;
        }
        cout << "--------------------------------\n" << endl;
    }
};

// ==========================================
// MAIN EXECUTION
// ==========================================
int main() {
    Graph g;

    // --- Building the Graph ---
    // Let's build a social network logic:
    // 0 is friends with 1 and 2
    // 1 is friends with 2
    // 2 is friends with 0 and 3
    /*
        0 --- 1
        |     |
        2 --- 3
         \   /
           4
    */

    g.addEdge(0, 1);
    g.addEdge(0, 2);
    g.addEdge(1, 2); // 1 and 2 are connected
    g.addEdge(2, 3);
    g.addEdge(3, 4);
    g.addEdge(2, 4);

    // Show the internal structure
    g.printGraph();

    // BFS: Should look wide first. 
    // From 0, it sees 1 and 2. Then it visits neighbors of 1, then neighbors of 2.
    // Expected: 0 1 2 3 4 (Order may vary slightly based on insertion)
    g.bfs(0);

    // DFS: Should go deep first.
    // From 0, go to 1. From 1, go to 2. From 2, go to 3...
    // Expected: 0 1 2 3 4 (But the path taken is different)
    g.dfs(0);

    return 0;
}

File: Java/Java/src/DataStructuresAndAlgorithms/Graphs.java
package JavaDSA;

import java.util.ArrayList;
import java.util.LinkedList;
import java.util.Queue;

/**
 * A Master Class demonstrating Graph Data Structures and Algorithms.
 * File Name: MasterGraph.java
 * * * * CONCEPTS:
 * 1. Adjacency Matrix: A 2D array where cell [i][j] = 1 means an edge exists.
 * - Space: O(V^2) -> Expensive!
 * - Lookup: O(1) -> Fast.
 * * 2. Adjacency List: An Array of Lists where index 'i' contains a list of neighbors.
 * - Space: O(V + E) -> Efficient.
 * - Lookup: O(degree) -> Slower than matrix.
 */
public class Graphs {

    public static void main(String[] args) {
        System.out.println("=== 1. ADJACENCY MATRIX (2D Array) ===");
        matrixDemo();

        System.out.println("\n=== 2. ADJACENCY LIST (ArrayList) ===");
        // We will use this graph for BFS and DFS
        // Graph Structure:
        // 0 -- 1 -- 2
        // |    |
        // 3 -- 4
        int V = 5; // Vertices 0 to 4
        ArrayList<ArrayList<Integer>> adj = new ArrayList<>();
        
        // Initialize empty lists for each vertex
        for (int i = 0; i < V; i++) {
            adj.add(new ArrayList<>());
        }

        // Add Edges (Undirected)
        addEdgeList(adj, 0, 1);
        addEdgeList(adj, 0, 3);
        addEdgeList(adj, 1, 2);
        addEdgeList(adj, 1, 4);
        addEdgeList(adj, 3, 4);

        printAdjList(adj);

        System.out.println("\n=== 3. BFS TRAVERSAL (Level Order) ===");
        bfs(adj, V, 0); // Start from node 0

        System.out.println("\n=== 4. DFS TRAVERSAL (Depth First) ===");
        dfs(adj, V, 0); // Start from node 0
    }

    // ==========================================================
    // SECTION 1: ADJACENCY MATRIX REPRESENTATION
    // ==========================================================
    public static void matrixDemo() {
        int v = 3;
        int[][] matrix = new int[v][v];

        // Edge 0-1
        matrix[0][1] = 1;
        matrix[1][0] = 1; // Undirected

        // Edge 1-2
        matrix[1][2] = 1;
        matrix[2][1] = 1;

        System.out.println("Matrix Representation:");
        for (int i = 0; i < v; i++) {
            for (int j = 0; j < v; j++) {
                System.out.print(matrix[i][j] + " ");
            }
            System.out.println();
        }
    }

    // ==========================================================
    // SECTION 2: ADJACENCY LIST HELPERS
    // ==========================================================
    public static void addEdgeList(ArrayList<ArrayList<Integer>> adj, int u, int v) {
        adj.get(u).add(v); // Add v to u's list
        adj.get(v).add(u); // Add u to v's list (Undirected)
    }

    public static void printAdjList(ArrayList<ArrayList<Integer>> adj) {
        for (int i = 0; i < adj.size(); i++) {
            System.out.print("Vertex " + i + ": ");
            for (int neighbor : adj.get(i)) {
                System.out.print(neighbor + " ");
            }
            System.out.println();
        }
    }

    // ==========================================================
    // SECTION 3: BFS (Breadth-First Search)
    // ==========================================================
    // Uses a QUEUE to explore neighbors level by level.
    public static void bfs(ArrayList<ArrayList<Integer>> adj, int V, int startNode) {
        boolean[] visited = new boolean[V];
        Queue<Integer> q = new LinkedList<>();

        visited[startNode] = true;
        q.add(startNode);

        System.out.print("BFS Path: ");
        while (!q.isEmpty()) {
            int curr = q.poll();
            System.out.print(curr + " ");

            // Get all neighbors of curr
            for (int neighbor : adj.get(curr)) {
                if (!visited[neighbor]) {
                    visited[neighbor] = true;
                    q.add(neighbor);
                }
            }
        }
        System.out.println();
    }

    // ==========================================================
    // SECTION 4: DFS (Depth-First Search)
    // ==========================================================
    // Uses RECURSION (System Stack) to go deep into one path.
    public static void dfs(ArrayList<ArrayList<Integer>> adj, int V, int startNode) {
        boolean[] visited = new boolean[V];
        System.out.print("DFS Path: ");
        dfsRecursive(adj, startNode, visited);
        System.out.println();
    }

    private static void dfsRecursive(ArrayList<ArrayList<Integer>> adj, int curr, boolean[] visited) {
        // Mark current node as visited and print it
        visited[curr] = true;
        System.out.print(curr + " ");

        // Recur for all connected neighbors
        for (int neighbor : adj.get(curr)) {
            if (!visited[neighbor]) {
                dfsRecursive(adj, neighbor, visited);
            }
        }
    }
}